### PR TITLE
Back end develop

### PR DESCRIPTION
**Pull request by MaayanBachar**

### CODE REVIEW

**Comment by MaayanBachar** [comment]
Note: please ignore generated rails files
Thanks!

File: db/schema.rb [hunk 1]
```diff
@@ -10,21 +10,31 @@
...
     t.string "address"
     t.integer "max_delivery_time", limit: 2
     t.string "image_url"
     t.datetime "created_at", null: false
     t.datetime "updated_at", null: false
+    t.boolean "tenbis"
   end
 
+  create_table "reviews", force: :cascade do |t|
+    t.bigint "restaurant_id"
```
Comment (GeReV) [inline db/schema.rb]: Check out `references`:
http://api.rubyonrails.org/v5.1/classes/ActiveRecord/ConnectionAdapters/TableDefinition.html#method-i-references

File: db/schema.rb [hunk 2]
```diff
@@ -10,21 +10,31 @@
...
   create_table "restaurants", force: :cascade do |t|
     t.string "name"
     t.string "cuisine"
-    t.integer "rating", limit: 2
-    t.integer "tenbis", limit: 2
     t.string "address"
     t.integer "max_delivery_time", limit: 2
     t.string "image_url"
     t.datetime "created_at", null: false
     t.datetime "updated_at", null: false
```
Comment (GeReV) [inline db/schema.rb]: You can use `t.timestamps` to generate these two.

File: db/schema.rb [hunk 3]
```diff
@@ -10,21 +10,31 @@
...
 
-ActiveRecord::Schema.define(version: 20180219073919) do
+ActiveRecord::Schema.define(version: 20180220100030) do
 
   # These are extensions that must be enabled in order to support this database
   enable_extension "plpgsql"
 
   create_table "restaurants", force: :cascade do |t|
     t.string "name"
     t.string "cuisine"
```
Comment (GeReV) [inline db/schema.rb]: Consider `null: false` in addition to `validates :whatever, presence: true`.

File: db/schema.rb [hunk 4]
```diff
@@ -10,21 +10,31 @@
...
+    t.bigint "restaurant_id"
+    t.string "reviewer_name"
+    t.integer "rating"
+    t.text "comment"
+    t.datetime "created_at", null: false
+    t.datetime "updated_at", null: false
+    t.index ["restaurant_id"], name: "index_reviews_on_restaurant_id"
+  end
+
+  add_foreign_key "reviews", "restaurants"
```
Comment (GeReV) [inline db/schema.rb]: What other table indices could you add here?

File: app/models/review.rb [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+#  rating        :integer
+#  comment       :text
+#  created_at    :datetime         not null
+#  updated_at    :datetime         not null
+#
+
+class Review < ApplicationRecord
+  belongs_to :restaurant
+
+  validates :reviewer_name, :presence => true
```
Comment (okeren-ww) [inline app/models/review.rb]: use the easier to write(and read) way - presence: true 
(it replaces the => way of writing)

File: app/models/restaurant.rb [hunk 6]
```diff
@@ -20,4 +42,26 @@ class Restaurant < ApplicationRecord
...
+        num_of_reviews += 1
+      end
+      total_rating / num_of_reviews
+    else
+      0
+    end
+
+
+
+  end
```
Comment (GeReV) [inline app/models/restaurant.rb]: I think you could have replaced this entire thing with `reviews.average(:rating)`.
(Provided by https://apidock.com/rails/ActiveRecord/Calculations/average)

Alternatively:
```ruby
reviews.map { |r| r.rating }.average

# or in shorthand:
reviews.map(&:rating).average
```
I think that is provided by [ActiveSupport](http://guides.rubyonrails.org/active_support_core_extensions.html).

You can also use `sum` and divide by `reviews.size`.
https://stackoverflow.com/questions/1341271/how-do-i-create-an-average-from-a-ruby-array

File: app/models/review.rb [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+#  updated_at    :datetime         not null
+#
+
+class Review < ApplicationRecord
+  belongs_to :restaurant
+
+  # change (added)
+  validates :reviewer_name, :presence => true
+  validates :rating, :presence => true
+  validates :comment, :presence => true
```
Comment (GeReV) [inline app/models/review.rb]: The slightly newer syntax would be `validates :rating, presence: true`

File: app/models/restaurant.rb (hunk 6 above)
Comment (GeReV) [reply app/models/restaurant.rb]: There are a lot of things you can do much more easily in Ruby and even more so in Rails.
It's worth searching for them.

File: db/migrate/20180220090641_change_tenbis_type_in_restaurants.rb [hunk 8]
```diff
@@ -0,0 +1,10 @@
+class ChangeTenbisTypeInRestaurants < ActiveRecord::Migration[5.1]
+  def change
+    remove_column :restaurants, :tenbis
+  end
+
+  # def self.down
```
Comment (okeren-ww) [inline db/migrate/20180220090641_change_tenbis_type_in_restaurants.rb]: I would remove comments of old code if you're not going to use it

File: app/models/restaurant.rb [hunk 9]
```diff
@@ -1,6 +1,28 @@
...
+  # A restaurant has many associated reviews.
+  # Each review contains a reference to its restaurant's ID
+  # The existence of reviews is dependent on the existence of the restaurant.
+  # has_many :reviews, dependent:  :destroy can be deleted
+  has_many :reviews
+
   # Validates that the text fields all contain something
-  validates :name, :cuisine, :rating, presence: true
+  validates :name, :cuisine, presence: true
   validates :tenbis, :address, :max_delivery_time, presence: true
```
Comment (GeReV) [inline app/models/restaurant.rb]: I'd suggest merging those:
```ruby
validates :name,
          :cuisine,
          :tenbis,
          ...,
          presence: true
```

File: app/models/review.rb (hunk 7 above)
Comment (GeReV) [reply app/models/review.rb]: You could also merge these together.

File: db/migrate/20180219130430_create_reviews.rb [hunk 10]
```diff
@@ -0,0 +1,12 @@
+class CreateReviews < ActiveRecord::Migration[5.1]
+  def change
+    create_table :reviews do |t|
+      t.belongs_to :restaurant, foreign_key: true
+      t.string :reviewer_name
+      t.integer :rating
```
Comment (GeReV) [inline db/migrate/20180219130430_create_reviews.rb:6]: Can this be `null`? Should it have a default value?

File: db/schema.rb [hunk 11]
```diff
@@ -10,21 +10,31 @@
...
   # These are extensions that must be enabled in order to support this database
   enable_extension "plpgsql"
 
   create_table "restaurants", force: :cascade do |t|
     t.string "name"
     t.string "cuisine"
-    t.integer "rating", limit: 2
-    t.integer "tenbis", limit: 2
     t.string "address"
     t.integer "max_delivery_time", limit: 2
```
Comment (GeReV) [inline db/schema.rb]: To be honest, I'm not even sure what `limit` means for `integer`...

**Review by GeReV — APPROVED** [review]

File: app/models/review.rb (hunk 5 above)
Comment (MaayanBachar) [reply app/models/review.rb]: Done. Thanks!

File: app/models/restaurant.rb (hunk 6 above)
Comment (MaayanBachar) [reply app/models/restaurant.rb]: Nice!! thanks!!!
changed to: 
reviews.size == 0 ? "-" : reviews.average(:rating).round

File: app/models/restaurant.rb [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+  validates :max_delivery_time, numericality: {greater_than_or_equal_to: 0}
+
+  # Validates image format
+  validates :image_url, allow_blank: true, format: {
+      with: %r{\.(gif|jpg|png)\Z}i,
+      message: 'must be a URL for GIF, JPG or PNG image.'
+  }
+
+  def get_rating
+    reviews.size == 0 ? "-" : reviews.average(:rating).round
```
Comment (okeren-ww) [inline app/models/restaurant.rb:41]: Great!

File: db/migrate/20180222122303_change_columns_in_reviews_cant_be_null.rb [hunk 13]
```diff
@@ -0,0 +1,7 @@
+class ChangeColumnsInReviewsCantBeNull < ActiveRecord::Migration[5.1]
+  def change
+    change_column_null(:reviews, :rating, false)
+    change_column_null(:reviews, :restaurant_id, false)
+    change_column_null(:reviews, :comment, false)
```
Comment (okeren-ww) [inline db/migrate/20180222122303_change_columns_in_reviews_cant_be_null.rb:5]: 👍

File: db/migrate/20180220100030_delete_rating_from_restaurants.rb [hunk 14]
```diff
@@ -0,0 +1,5 @@
+class DeleteRatingFromRestaurants < ActiveRecord::Migration[5.1]
+  def change
+    remove_column :restaurants, :rating
```
Comment (okeren-ww) [inline db/migrate/20180220100030_delete_rating_from_restaurants.rb:3]: Excellent

**Review by okeren-ww — APPROVED** [review]
Excellent!

File: app/models/restaurant.rb (hunk 12 above)
Comment (GeReV) [reply app/models/restaurant.rb:41]: +1!

**Review by GeReV — APPROVED** [review]
Looks great!

**Pull request merged**